### PR TITLE
update OTLP to v0.8.0

### LINKED ISSUE #3534: Update Collector to OTLP 0.8.0
This is the umbrella issue which will have several items detailing tasks that need to be completed to update the Collector to OTLP version 0.8.0. Stay tuned.

https://github.com/open-telemetry/opentelemetry-proto/releases/tag/v0.8.0

Roadmap:

- [x] Implement all naming changes in model/pdata/metrics.
- [x] Upgrade proto submodule to 0.8.0 without doing any translation. This may require changing pdata since some protos changed the value to oneof.
  - [x] If change to large, split into multiple upgrades to intermediate steps.
- [x] Implement required translation (as an example https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L514).
- [x] Re-enable staticcheck in pdata/metrics*.go
- [x] Remove primitiveAsDoubleField

### PR DESCRIPTION
**Pull request by codeboten**
**Description:**
Updating OTLP to version 0.8.0. Here is a checklist of the progress so far:
- [x] bump submodule version
- [x] regenerate protos
- [x] update uses of `DoubleGauge`, `DoubleSum`, `DoubleHistogram`, `DoubleSummary`, `DoubleDataPoint`, `DoubleHistogramDataPoint`, `DoubleSummaryDataPoint`, and `DoubleExemplar`
- [x] update `ExemplarSlice` to use `[]*otlpmetrics.Exemplar`
- [ ] deprecate `IntSum`, `IntGauge` https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L156
- [ ] deprecate `IntDataPoint` https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L554
- [ ] deprecate `IntExemplar` https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L675
- [ ] deprecate `IntHistogram` https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L170
- [ ] deprecate filtered labels https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L514

**Link to tracking Issue:** Part of #3534

### CODE REVIEW

**Comment by codeboten** [comment]
I've started the work to upgrade OTLP here, but I would like some feedback from @open-telemetry/collector-approvers on the way to proceed forward. The linter is currently failing because of various deprecated types still in use:
```
pdata/metrics.go:193:58: SA1019: otlpmetrics.IntGauge is deprecated: Do not use. (staticcheck)
                ms.orig.Data = &otlpmetrics.Metric_IntGauge{IntGauge: &otlpmetrics.IntGauge{}}
                                                                       ^
pdata/metrics.go:197:54: SA1019: otlpmetrics.IntSum is deprecated: Do not use. (staticcheck)
                ms.orig.Data = &otlpmetrics.Metric_IntSum{IntSum: &otlpmetrics.IntSum{}}
                                                                   ^
pdata/metrics.go:201:66: SA1019: otlpmetrics.IntHistogram is deprecated: Do not use. (staticcheck)
                ms.orig.Data = &otlpmetrics.Metric_IntHistogram{IntHistogram: &otlpmetrics.IntHistogram{}}
                                                                               ^
pdata/metrics.go:261:51: SA1019: otlpmetrics.IntGauge is deprecated: Do not use. (staticcheck)
                data := &otlpmetrics.Metric_IntGauge{IntGauge: &otlpmetrics.IntGauge{}}
                                                                ^
pdata/metrics.go:269:47: SA1019: otlpmetrics.IntSum is deprecated: Do not use. (staticcheck)
                data := &otlpmetrics.Metric_IntSum{IntSum: &otlpmetrics.IntSum{}}
                                                            ^
pdata/metrics.go:277:59: SA1019: otlpmetrics.IntHistogram is deprecated: Do not use. (staticcheck)
                data := &otlpmetrics.Metric_IntHistogram{IntHistogram: &otlpmetrics.IntHistogram{}}
                                                                        ^
pdata/metrics_test.go:47:17: SA1019: otlpmetrics.IntGauge is deprecated: Do not use. (staticcheck)
                                        IntGauge: &otlpmetrics.IntGauge{},
                                                   ^
pdata/metrics_test.go:63:15: SA1019: otlpmetrics.IntSum is deprecated: Do not use. (staticcheck)
                                        IntSum: &otlpmetrics.IntSum{},
                                                 ^
pdata/metrics_test.go:79:21: SA1019: otlpmetrics.IntHistogram is deprecated: Do not use. (staticcheck)
                                        IntHistogram: &otlpmetrics.IntHistogram{},
                                                       ^
pdata/metrics_test.go:441:26: SA1019: otlpmetrics.IntDataPoint is deprecated: Do not use. (staticcheck)
                                                                                DataPoints: []*otlpmetrics.IntDataPoint{
                                                                                               ^
pdata/metrics_test.go:776:20: SA1019: otlpmetrics.IntDataPoint is deprecated: Do not use. (staticcheck)
                                DataPoints: []*otlpmetrics.IntDataPoint{
```

Would it be better for me to continue working through those messages in this PR or disable the linter for `SA1019` to allow this code to merge and prevent the PR from getting too large? Happy to go either ways here.

**Comment by tigrannajaryan** [comment]
Thanks for starting this @codeboten !

We can probably silence the linter temporarily.

However, I think we need a more elaborate plan for this change.

Changing this repo is relatively easy. I think contrib is going to be much bigger volume of changes. We need to time the change such that it is fully doable in the period between the releases (2 weeks). Perhaps once this PR is ready, don't merge it yet, prepare the PR for corresponding contrib changes and only after ensuring contrib is also ready to be merged move forward. Contrib changes may turn out to be a lot of work and we may need to ask codeowners of each component to prepare a PR for their own component (or maybe not if it turns out to be not as laborious as I am imagining).

**Comment by codeboten** [comment]
I tested the contrib repo with the changes as they are in this PR currently and things worked, this is expected as I'm not yet renaming things in pdata. We could potentially move ahead with this PR, which gives us at least the new protos and is not too hard to review.

The remaining items in the checklist could be tackled in following PRs and potentially even done in parallel. Thoughts @tigrannajaryan & @bogdandrutu?

File: .golangci.yml
Comment (bogdandrutu) [inline .golangci.yml:138]: We should create an issue, or document this in #3534 that is left to deal with for later.

**Review by bogdandrutu — COMMENTED** [review]
I would not "fix" #3534 since this change indeed does the upgrade, but does not do the right transformations to/from deprecated fields/metric types.

**Comment by tigrannajaryan** [comment]
Reading the proto changes in 0.8.0 it seems like the wire format changes are not breaking? From what I see it should remain wire-format compatible with previous version (fields in some messages are added but not removed and some messages are renamed - all of which is backward compatible). Is that correct or I am missing a breaking change?

File: model/pdata/generated_metrics.go
Comment (tigrannajaryan) [inline model/pdata/generated_metrics.go:1194]: What's the future plan? Will this struct be renamed `NumberDataPoint`?

File: model/pdata/generated_metrics.go
Comment (tigrannajaryan) [inline model/pdata/generated_metrics.go:1237]: Do we plan to remove `Value` and expose 2 different types of values, the `int` and the `double` in a future PR?

Comment (bogdandrutu) [reply model/pdata/generated_metrics.go:1237]: Yes

**Review by bogdandrutu — APPROVED** [review]
I think we can merge this and do the rest of the work in followup PRs

File: model/pdata/generated_metrics.go
Comment (codeboten) [reply model/pdata/generated_metrics.go:1194]: I plan on renaming this next and tackling removing IntDataPoint at the same time.

**Pull request merged**
